### PR TITLE
Versioning  active model

### PR DESCRIPTION
**Pull request by matDobek**
# Ticket
https://netguru.atlassian.net/browse/PEOPLE-128

# Description
Versioning via ActiveModel based on: http://strikingly.github.io/blog/2015/09/14/Simple-rails-history-pattern-ActiveRecord/
This is only backend solution.

At beginning this solutions seems to be better, than using gem, as:
* we _don't_ have to remember to update serialized data, when column names are changed / added. Gems provide versioning via serializing model into one field - so if we plan to change column name, or add new columns, we also have to update all already serialized columns
* Easy manipulate data on SQL level - operations like grouping by date, can be done in SQL, when we keep old records in separate tables, instead of serializing them

However, right now it feels like a Goldberg machine
image

It provides unnecessary complexity, to a simple task, just to be able to handle corner cases, a bit more efficient.

### Before merge checklist
- [x] CircleCI is passing
- [ ] Proper labels are set

### CODE REVIEW

**Comment by DariuszMusielak** [comment]
We will use something similar to this one.

File: app/validators/user_skill_rate/rate_within_range_validator.rb [hunk 1]
```diff
@@ -11,7 +11,7 @@ def validate(record)
     attr_reader :record
 
     def rate_within_range?
-      rate_type = record.skill&.rate_type
+      rate_type = record.user_skill_rate&.skill&.rate_type
```
Comment (netguru-hound) [inline app/validators/user_skill_rate/rate_within_range_validator.rb]: unexpected token error

File: app/models/user_skill_rate/content.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+class UserSkillRate::Content < ActiveRecord::Base
```
Comment (netguru-hound) [inline app/models/user_skill_rate/content.rb]: Use nested module/class definitions instead of compact style.

File: app/models/user.rb [hunk 3]
```diff
@@ -172,15 +172,17 @@ def user_memberships_repository
...
       .where('user_skill_rates.user_id' => self.id)
-      .where('user_skill_rates.rate > 0')
+      .where('user_skill_rate_contents.rate > 0')
   end
 
   def non_rated_skills
     ::Skill
-      .joins(:user_skill_rates)
+      .joins(user_skill_rates: :contents)
+      .where('user_skill_rate_contents.created_at = (SELECT MAX(user_skill_rate_contents.created_at) FROM user_skill_rate_contents WHERE user_skill_rates.id = user_skill_rate_contents.user_skill_rate_id)')
```
Comment (netguru-hound) [inline app/models/user.rb]: Line is too long. [205/100]

File: app/models/user.rb [hunk 4]
```diff
@@ -172,15 +172,17 @@ def user_memberships_repository
 
   def rated_skills
     ::Skill
-      .joins(:user_skill_rates)
+      .joins(user_skill_rates: :contents)
+      .where('user_skill_rate_contents.created_at = (SELECT MAX(user_skill_rate_contents.created_at) FROM user_skill_rate_contents WHERE user_skill_rates.id = user_skill_rate_contents.user_skill_rate_id)')
```
Comment (netguru-hound) [inline app/models/user.rb]: Line is too long. [205/100]

File: app/services/skills/user_skill_rates/create.rb [hunk 5]
```diff
@@ -1,21 +1,26 @@
...
 
       private
 
-      attr_reader :user_id, :skill_id, :rate
+      def user_skill_rate
+        @user_skill_rate ||= ::UserSkillRate.create(user_id: user_id, skill_id: skill_id)
+      end
+
+      def user_skill_rate_content_params
+        user_skill_rate.attributes.slice('rate', 'note', 'favorite')
```
Comment (matDobek) [inline app/services/skills/user_skill_rates/create.rb]: ❤️

**Comment by matDobek** [comment]
by far, lgtm 👍  
( except specs :) )

File: spec/services/skills/user_skill_rates/update_spec.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    it 'updates favorite on user rate skill' do
+      expect{ subject.call }.to change{ user_skill_rate.reload.rate }.from(0).to(1)
+    end
+
+    it 'updates favorite on user favorite skill' do
+      expect{ subject.call }.to change{ user_skill_rate.reload.favorite }.from(false).to(true)
+    end
+
+    it 'creates new user_skill_rate_content' do
+      expect{ subject.call }.to change{ user_skill_rate.contents.count }
```
Comment (netguru-hound) [inline spec/services/skills/user_skill_rates/update_spec.rb]: Space missing to the left of {.

File: spec/services/skills/user_skill_rates/update_spec.rb [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+      {
+        id: user_skill_rate.id,
+        note: 'def',
+        rate: 1,
+        favorite: true
+      }
+    }
+
+    it 'updates note on user rate skill' do
+      expect{ subject.call }.to change{ user_skill_rate.reload.note }.from('abc').to('def')
```
Comment (netguru-hound) [inline spec/services/skills/user_skill_rates/update_spec.rb]: Space missing to the left of {.

File: spec/services/skills/user_skill_rates/update_spec.rb [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    )
+  end
+
+  describe '#call' do
+    let(:last_content) { user_skill_rate.contents.last }
+    let(:user) { create(:user) }
+    let(:user_skill_rate) do
+      create(:user_skill_rate, note: 'abc', rate: 0, favorite: false, user: user)
+    end
+    let(:params) {
```
Comment (netguru-hound) [inline spec/services/skills/user_skill_rates/update_spec.rb]: Avoid using {...} for multi-line blocks.

File: spec/services/skills/user_skill_rates/create_spec.rb [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    )
+  end
+  let(:user) { create(:user) }
+  let(:skill) { create(:skill) }
+
+  describe '#call' do
+    let(:user_skill_rates) { user.reload.user_skill_rates }
+
+    it 'creates UserSkillRate' do
+      expect{ subject.call }.to change{ user_skill_rates.count }
```
Comment (netguru-hound) [inline spec/services/skills/user_skill_rates/create_spec.rb]: Space missing to the left of {.

File: spec/controllers/user_skill_rates_controller_spec.rb [hunk 10]
```diff
@@ -23,10 +23,12 @@
...
       subject
-      expect(controller.grouped_skills_by_category.values.count).to eq(2)
-      expect(controller.grouped_skills_by_category.keys).to include('backend', 'frontend')
-      expect(controller.grouped_skills_by_category['backend']).to include(user_skill_rate_backend)
-      expect(controller.grouped_skills_by_category['frontend']).to include(user_skill_rate_frontend)
+      aggregate_failures do
+        expect(controller.grouped_skills_by_category.values.count).to eq(2)
+        expect(controller.grouped_skills_by_category.keys).to include('backend', 'frontend')
+        expect(controller.grouped_skills_by_category['backend']).to include(user_skill_rate_backend)
+        expect(controller.grouped_skills_by_category['frontend']).to include(user_skill_rate_frontend)
```
Comment (netguru-hound) [inline spec/controllers/user_skill_rates_controller_spec.rb]: Line is too long. [102/100]

**Pull request merged**